### PR TITLE
WIP - Packaging

### LINKED ISSUE #16: Need installer binary
not high priority, but we eventually need a single binary that will install ipfs and the electron shell app.

### LINKED ISSUE #20: windows installer
as a followup to #16, we can use https://github.com/atom/grunt-electron-installer to build a windows installer for us, it looks pretty nice and should help a lot with getting people to have ipfs installed.

### LINKED ISSUE #508: Add a package builder that generates installers for Win, Mac and Linux
Everyday users want to use a native OS installer to install apps. In order to support this, we need to use a package builder like [electron-builder](https://www.npmjs.com/package/electron-builder) to generate those installers.

What to do:
- [ ] Add an electron package builder like [electron-builder](https://www.npmjs.com/package/electron-builder) that gan generate installers for Windows, Mac and Linux
- [ ] Configure the builder so we can automatically generate new installers and release them to dist.ipfs.io 

There's some helpful, but slightly outdated, info in #20.

Related Issues: closes #16, closes #20

### PR DESCRIPTION
**Pull request by VictorBjelkholm**
OSX build seems to work fine, ends up with a `.dmg at least`.

Linux fails on RPM builds. 

Example artifacts: https://ci.ipfs.team/blue/organizations/jenkins/IPFS%20Shipyard%2Fstation/detail/packaging/5/artifacts/

Missing waiting for windows. Will continue with this

----

Close #508, close #20 and close #16.

### CODE REVIEW

**Comment by hacdias** [comment]
**To close on merge:**

- #508
- #48
- #20
- #16

**Comment by diasdavid** [comment]
@VictorBjelkholm @hacdias what's missing to make this happen?

**Comment by VictorBjelkholm** [comment]
@diasdavid verifying that the osx build actually works. It successfully built but has not been verified by anyone yet.

Then getting the builds for linux and windows to produce artifacts, working on this atm.

**Comment by VictorBjelkholm** [comment]
Got all the builds to successfully turn into a package on all OSes. Left is to verify that they actually work. Also, would like to change the version from being the version set in `package.json` to the current commit instead.

**Comment by VictorBjelkholm** [comment]
https://github.com/ipfs-shipyard/station/commit/8b3a6fcbc0a6d8e2320b6a44eb77ee146163b286#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L13 screwed me a bit after getting all the builds working and then doing a rebase from master. Fixed this and waiting for build to finish now.

**Comment by VictorBjelkholm** [comment]
.deb on Ubuntu 16.04.3 is acting weird and not working. After installing, it doesn't add anything to the dash menu or in $PATH, leading to not being able to open the application unless manually executing `/usr/lib/ipfs-station/IPFS\ Station` which `/usr/bin/ipfs-station` is supposed to be a symbolic link to, but it points to `../lib/ipfs-station/ipfs-station` instead (notice the whitespace and mix lower/upper case).

When manually running the right executable, I get the following error:

```
$ /usr/lib/ipfs-station/IPFS\ Station
A JavaScript error occurred in the main process
Uncaught Exception:
Error: Cannot find module 'electron-compile'
    at Module._resolveFilename (module.js:470:15)
    at Function.Module._resolveFilename (/usr/lib/ipfs-station/resources/electron.asar/common/reset-search-paths.js:35:12)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/usr/lib/ipfs-station/resources/app/es6-shim.js:4:23)
    at Object.<anonymous> (/usr/lib/ipfs-station/resources/app/es6-shim.js:10:3)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
```

So at least two things are not working when installing/packaging for .deb

**Comment by VictorBjelkholm** [comment]
Did a quick test on windows and getting the same error as on Ubuntu. Not sure what's wrong in the packaging process.

**Comment by VictorBjelkholm** [comment]
These packages should work better. **Please try them and report any issues!**


- DMG for macOS: https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/PR-514/18/artifact/out/make/IPFS%20Station-0.1.0.dmg
- RPM for Red Hat et al: https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/PR-514/18/artifact/out/make/ipfs-station-0.1.0.x86_64.rpm
- DEB for Debian + Ubuntu: https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/PR-514/18/artifact/out/make/ipfs-station_0.1.0_amd64.deb
- EXE for Windows: https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/PR-514/18/artifact/out/make/squirrel.windows/x64/IPFS%20Station-0.1.0%20Setup.exe
- NUPKG for Windows (as well?): https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/PR-514/18/artifact/out/make/squirrel.windows/x64/Station-0.1.0-full.nupkg

**Comment by VictorBjelkholm** [comment]
Everything working OK on macOS now, more testing would be welcome though!

Ubuntu version is still acting up, now getting the following error:

```
non-zero exit code 1

            while running: /usr/lib/ipfs-station/resources/app/node_modules/go-ipfs-dep/go-ipfs/ipfs daemon


            internal/child_process.js:319
    throw errnoException(err, 'spawn');
    ^

Error: spawn EACCES
    at exports._errnoException (util.js:1050:11)
    at ChildProcess.spawn (internal/child_process.js:319:11)
    at exports.spawn (child_process.js:390:9)
    at module.exports (/usr/lib/ipfs-station/resources/app/node_modules/comandante/index.js:11:14)
    at Object.<anonymous> (/usr/lib/ipfs-station/resources/app/node_modules/subcomandante/subcom:23:13)
    at Object.<anonymous> (/usr/lib/ipfs-station/resources/app/node_modules/subcomandante/subcom:118:3)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
```

Seems the `ipfs` binary from `go-ipfs-dep` is installed with only `root` having access to it, so the normal user cannot run it. If I run `sudo chown user /usr/lib/ipfs-station/resources/app/node_modules/go-ipfs-dep/go-ipfs/ipfs` and after try to run Station again, everything works OK.

Windows version pops up some weird code editor animation for a minute or something like that (probably want to replace with the IPFS logo or similar before proper release [also has the default winforms icon in the task bar])  which then just disappears without any notice, and now station is installed? Anyways, once installed, it doesn't seem to transition from the loading screen to the screen where I can actually see my node.

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "file-extension": "^4.0.0",
     "ipfs-api": "^17.1.3",
     "ipfs-geoip": "^2.3.0",
+    "ipfs-logo": "^0.2.0",
```
Comment (diasdavid) [inline package.json]: You should not need this anymore. Make sure to rebase master onto this branch @VictorBjelkholm

**Comment by hacdias** [comment]
The installer above works for me on windows, but the icons are broken. Please rebase into master. I also have to update the icons on package.json since they're still using ipfs-logo.

**Comment by machawk1** [comment]
I poked around the macOS version linked from the dmg above. Based on my limited experience with running it from Terminal, the behavior appears consistent.

Is there a plan to use a station-specific icon for the .app or to retain the IPFS cube?

**Comment by hacdias** [comment]
I'll be a bit busy this week and the next week, but I'll take a look at [electron-builder](https://github.com/electron-userland/electron-builder). It seems a bit more flexible and widely known. Let me test it today @VictorBjelkholm and then I'll leave an update here. If I need to change something, I'll use this branch, okay?

/cc @diasdavid

**Comment by hacdias** [comment]
> Windows version pops up some weird code editor animation for a minute or something like that (probably want to replace with the IPFS logo or similar before proper release [also has the default winforms icon in the task bar]) which then just disappears without any notice, and now station is installed? Anyways, once installed, it doesn't seem to transition from the loading screen to the screen where I can actually see my node.

ElectronForge uses Squirrel installers for windows whose slogan is "Squirrel: It's like ClickOnce but Works™".  I think you can take your conclusions about "disappears without any notice" 😄

**Comment by hacdias** [comment]
I opened an [issue](https://github.com/electron-userland/electron-builder/issues/2361) on electron-builder to check if they support electron-prebuild-compile. If so, I think electron-builder would be a better option to use since it's more flexible.

**Comment by diasdavid** [comment]
@hacdias let's avoid using the word _cosmos_ as it might confuse users with another P2P Network. Try perhaps "Becoming InterPlanetary"

**Comment by VictorBjelkholm** [comment]
Jenkins is now also setting the status of the commit, where you'll always be able to find the built binaries for the three platforms. 

image

**Comment by hacdias** [comment]
@VictorBjelkholm [jenkins](https://ci.ipfs.team/job/IPFS%20Shipyard/job/station/job/packaging/) says the last build was 7 days ago. Am I looking in the wrong place?

**Comment by VictorBjelkholm** [comment]
@hacdias yes, that's the wrong place. Correct latest build associated with this PR is in the status check for easy access. Just click on "Show all checks", then click on "Details" to the right of the "continuous-integration/jenkins/pr-merge" job. You should be taken to https://ci.ipfs.team/blue/organizations/jenkins/IPFS%20Shipyard%2Fstation/detail/PR-514/22/pipeline then which is the latest job. Shows up as "3 hours ago" for me.

**Comment by hacdias** [comment]
Strangely, it is building binaries with ~120MB for windows with the old version.

**Comment by hacdias** [comment]
We should also take a look [at this](https://github.com/Squirrel/Squirrel.Windows/blob/master/docs/using/application-signing.md) since Windows executables must be signed in order not to blocked by antivirus and/or windows firewall.

**Comment by hacdias** [comment]
@VictorBjelkholm I think the installers are not working since it is packaging the old version somehow.

**Comment by diasdavid** [comment]
What about merging this PR and working from master? So that you are sure that you are working in the latest

**Comment by hacdias** [comment]
Yeah, I agree @diasdavid

**Pull request merged**